main.rb

```diff
@@ -40,116 +40,134 @@ def my_select
-    elsif param.is_a? Class
-      each { |item| return false if item.instance_of? param }
-    else
-      each { |item| return false if item == param }
+      return true
+    end
+    state = case param
+            when Regexp then check_regexp(false, param)
+            when Class then check_class(false, param)
+            else check(false, param)
+            end
+    state == false ? state : true
+  end
+
+  def check_regexp(status, param, negative = false)
+    each { |item| return status unless param.match(item) } if negative
+    each { |item| return status if param.match(item) } unless negative
+  end
+
+  def check_class(status, param, negative = false)
+    each { |item| return status unless item.instance_of? param } if negative
+    each { |item| return status if item.instance_of? param } unless negative
+  end
+
+  def check(status, param, negative = false)
+    each do |item|
+      cond = if param == true
+               !item.nil? && item
+             else
+               item == param
+             end
+
+      cond = negative ? !cond : cond
+
+      return status if cond
     end
-    true
   end
 
   def my_count(cond = nil)
     counter = 0
     each do |item|
-      counter += 1 if (block_given? && yield(item)) || (item == cond) || !cond
+      counter += 1 if (block_given? && yield(item)) || (item == cond) || (!cond && !block_given?)
     end
     counter
   end
 
   def my_inject(*args)
-    acc = 0
-    acc = args[0] if !args[0].is_a?(Symbol) && !args.empty?
-
-    return multiply_els(self) if args.include? :*
+    acc = setup_acc(args)
+    return multiply_els(self, acc) if args.include? :*
 
-    return add_els(self) if args.include? :+
+    return add_els(self, acc) if args.include? :+
 
-    return divide_els(self) if args.include? :/
+    return divide_els(self, acc) if args.include? :/
 
-    return subs_els(self) if args.include? :-
+    return subs_els(self, acc) if args.include? :-
 
+    each { |item| acc = yield(acc, item) }
+    acc
+  end
 
-    each do |item|
-      acc = yield(acc, item)
+  def setup_acc(param)
+    if !param[0].is_a?(Symbol) && !param.empty?
+      param[0]
+    elsif param.my_all?(String) && param[0].is_a?(String) then param[0]
+    elsif param.my_all?(String) then ''
+    else 0
     end
-    acc
   end
 
-  def multiply_els(arr)
-    arr.my_inject(1) { |acc, item| acc * item }
+  def multiply_els(arr, accum)
+    accum = accum.zero? ? 1 : accum
+    arr.my_inject(accum) { |acc, item| acc * item }
   end
 
-  def add_els(arr)
-    arr.my_inject(0) { |acc, item| acc + item }
+  def add_els(arr, accum)
+    arr.my_inject(accum) { |acc, item| acc + item }
   end
 
-  def divide_els(arr)
-    arr.my_inject(1.0) { |acc, item| acc / item }
+  def divide_els(arr, accum)
+    arr.my_inject(accum.to_f) { |acc, item| acc / item }
   end
 
-  def subs_els(arr)
-    arr.my_inject(0) { |acc, item| acc - item }
+  def subs_els(arr, accum)
+    arr.my_inject(accum) { |acc, item| acc - item }
   end
```

Comment:
Let's move `multiply_els(arr)` and other supporting methods outside this module. This would make us call it outside this way - `multiply_els([1,2,3])`.